### ISSUE TITLE
`index` following destination tensor's stride

### ISSUE THREAD

**Issue by nhynes** [torch/torch7#725]
Often, when loading data, there's a need to place a tensor within a larger tensor (e.g., when adding padding/end tokens). The simple solution is to just have a loop in which the appropriate slicing occurs but it would be more convenient to be able to do a single `:index`. Unfortunately, `THTensor_(indexSelect)` trashes the stride information of the dest tensor which generally makes this impossible.

As an example:
```lua
local t = torch.rand(4, 3)
local u = torch.zeros(2, 6)
local inds = torch.LongTensor{1, 3}
torch.Tensor():set(u:storage(), 2, 4, 5, 3, 1)\index(t, 1, inds)
```
One might expect rows 1 and 3 of `t` to be centered within `u`, but this actually produces the selected elements of `t` moving along `u` as if the stride were 1 since the operation is just a [memcpy](https://github.com/torch/torch7/blob/master/lib/TH/generic/THTensorMath.c#L167).

It's probably much too late for this, but it might be nice if `index` could use the destination tensor's stride. I'd be happy to take a whack at this if it's not out of the question!

**Comment by soumith** [comment]
Would love it if you took a whack at it. I dont think it's unreasonable to change this.

**Comment by nhynes** [comment]
Actually, I think that I may have had a bug in my example. The destination tensor shouldn't be contiguous so the memcpy branch shouldn't be taken!

If anyone else stumbles upon this issue, though, a corrected (and working) example is
```lua
local t = torch.rand(4, 3, 5)
local u = torch.zeros(4, 5, 7)
local inds = torch.LongTensor{1, 3}
local fullstep = u:stride(1) + u:stride(2) + u:stride(3)

torch.Tensor():set(u:storage(), fullstep+1,
  2, u:stride(1),
  t:size(2), u:stride(2),
  t:size(3), u:stride(3)
):index(t, 1, inds)

print(u)
```

Sorry for the confusion!
